Samplers + Rcpp Interface (Milestone II and III)

papachristoumarios opened the pull request:
This pull request contains the following contributions

 * Hamiltonian Monte Carlo sampling with reflections 
    * Leapfrog integrator of [1]
   *  Collocation method of [2] 
 * Underdamped Langevin Diffusion with the Randomized Midpoint Method of [3]
 * Bug-fixes on ODE solvers
 * Refactor ODE solvers to work with functors instead of `std::function`
 * ODE solver tests in higher dimensions
 * Rcpp bindings
   * R function pointer wrapper functors
   * Functionality of `sample_points`
   * Functionality of `ode_solve`
 
A [tutorial](https://papachristoumarios.github.io/2020/07/21/Sampling-from-high-dimensional-truncated-log-concave-densities-with-volesti/) is currently being drafted.

## References

1. Lee, Yin Tat, Ruoqi Shen, and Kevin Tian. "Logsmooth Gradient Concentration and Tighter Runtimes for Metropolized Hamiltonian Monte Carlo." arXiv preprint arXiv:2002.04121 (2020).
2. Lee, Yin Tat, Zhao Song, and Santosh S. Vempala. "Algorithmic theory of ODEs and sampling from well-conditioned logconcave densities." arXiv preprint arXiv:1812.06243 (2018).
3. Shen, Ruoqi, and Yin Tat Lee. "The randomized midpoint method for log-concave sampling." Advances in Neural Information Processing Systems. 2019.

vissarion commented on R-proj/src/ode_solve.cpp:
```diff
@@ -0,0 +1,202 @@
+
+// VolEsti (volume computation and sampling library)
+
+// Copyright (c) 20012-2020 Vissarion Fisikopoulos
```
20012 is too late for humanity ;)

vissarion commented on include/random_walks/hamiltonian_monte_carlo_walk.hpp:
```diff
@@ -0,0 +1,168 @@
+    inline NT hamiltonian(Point &pos, Point &vel) const {
+      return f(pos) + 0.5 * vel.dot(vel);
+    }
+    //
```
if not needed please remove

vissarion commented on include/random_walks/hamiltonian_monte_carlo_walk.hpp:
```diff
@@ -0,0 +1,168 @@
+      NT epsilon_=1e-4)
+    {
+      epsilon = epsilon_;
+      eta = 1.0 / (sqrt(pow(dim, 3) * F.params.L));
```
it would be useful to link the publication that provides an evidence for using this bound

vissarion commented on include/ode_solvers/randomized_midpoint.hpp:
```diff
@@ -0,0 +1,226 @@
+
+	void calculate_Ws(NT &a, RandomNumberGenerator &rng) {
+		// Initialize matrices to zero
+		Y = 0 * Y;
```
too much indentation

vissarion commented on include/ode_solvers/randomized_midpoint.hpp:
```diff
@@ -0,0 +1,226 @@
+      }
+      else {
+        // Find intersection (assuming a line trajectory) between x and y
+				do {
```
weird indentation

vissarion commented on include/ode_solvers/randomized_midpoint.hpp:
```diff
@@ -0,0 +1,226 @@
+
+// Based on https://papers.nips.cc/paper/8483-the-randomized-midpoint-method-for-log-concave-sampling.pdf
+
+#ifndef RANDOMIZED_MIDPOINT_HPP
```
there is an indentation issue in the whole file

vissarion commented on include/ode_solvers/oracle_functors_rcpp.hpp:
```diff
@@ -0,0 +1,120 @@
+#define ORACLE_FUNCTORS_RCPP_HPP
+
+enum ode_solvers {
+  no_solver,
```
what happens in the `no_solver` case ? `rk` could have been `range-kutta`

papachristoumarios replied:
The no_solver case takes care of the case that we do uniform sampling. 
the elements of this enum are used for the sample_points() function of the R API
and the ode_solve function. We want from parsing the user arguments supplied
to sample_points to do a switch regarding which solver to use.

papachristoumarios replied:
Also changed `rk` to `runge-kutta`

vissarion commented on include/ode_solvers/leapfrog.hpp:
```diff
@@ -24,65 +29,78 @@ class LeapfrogODESolver {
-            xs[x_index] += y;
-          }
+            xs[x_index] = xs_prev[x_index] + y;
+					}
```
something is fishy here...

vissarion commented on include/ode_solvers/leapfrog.hpp:
```diff
@@ -11,11 +11,16 @@
 #ifndef LEAPFROG_HPP
```
there is also an issue with indentation in this file

vissarion commented on include/ode_solvers/integral_collocation.hpp:
```diff
@@ -0,0 +1,281 @@
+      }
+
+    }
+
```
too many blank lines

vissarion commented on include/generators/boost_random_number_generator.hpp:
```diff
@@ -61,7 +59,7 @@ private :
 struct BoostRandomNumberGenerator<RNGType, NT, Seed>
 {
-    BoostRandomNumberGenerator(int d)
+    BoostRandomNumberGenerator(int d=1)
```
why do you need a default value?

papachristoumarios replied:
Ok. I restored the default one.

vissarion commented on include/convex_bodies/hpolytope.h:
```diff
@@ -99,20 +99,20 @@ class HPolytope {
 
             NT const tol = 0.00000001;
             std::tuple<VT, NT, bool> inner_ball = max_inscribed_ball(A, b, 150, tol);
-        
```
are those changes by mistake?

papachristoumarios replied:
I think this is due to another branch that has merged these changes into develop.

vissarion commented on examples/logconcave/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+# VolEsti (volume computation and sampling library)
+# Copyright (c) 20012-2018 Vissarion Fisikopoulos
```
20012 --> 2012

vissarion reviewed (approved):
Thanks @papachristoumarios for this great PR! I am OK with merging, I have a few comments. Additionally:
-- Binary files (e.g. png files and executables in examples) should be removed
-- Adding examples for other implemented functions could be beneficial e.g. langevin walk sampling and SDE solver

EDIT: also please change the CMakeList.txt in `examples` to be independent in each directory i.e. log-concave examples should not need lapack and blas to build

papachristoumarios commented:
Hi @vissarion,

The changes you have requested have been carried out. 
Moreover, the png files in the examples folder serve as graphics for the corresponding readme,
i.e. examples/logconcave/README.md. 

I have also added an example for ULD (Langevin Walk) for the same density in an untruncated setting
(works in the truncated one as well) and the SDE (Randomized Midpoint), and modified CMakeLists such
that the examples can be built independently.

vissarion commented on examples/logconcave/CMakeLists.txt:
```diff
@@ -0,0 +1,113 @@
+
+  add_executable (simple_hmc simple_hmc.cpp)
+  add_executable (simple_ode simple_ode.cpp)
+  add_executable (simple_uld simple_uld.cpp)
```
this file seems to be missing

papachristoumarios committed:
Squash commits into one

papachristoumarios commented:
Discussion moved to #124
